vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+	    ec.l.multilabels = multilabels;
+
+	    return hamming_loss(ec.l.multilabels.label_v, selected_labs);
+    }
+
+
+    //pick up the "closest" example in the leaf using the score function.
+    int64_t pick_nearest(memory_tree& b, single_learner& base, const uint32_t cn, example& ec)
+    {
+        if (b.nodes[cn].examples_index.size() > 0)
+        {
+            float max_score = -FLT_MAX;
+            int64_t max_pos = -1;
+            for(size_t i = 0; i < b.nodes[cn].examples_index.size(); i++)
+            {
+                float score = 0.f;
+                uint32_t loc = b.nodes[cn].examples_index[i];
+
+                //do not use reward to update memory tree during the very first pass 
+                //(which is for unsupervised training for memory tree)
+                if (b.learn_at_leaf == true && b.current_pass >= 1){ 
+                    float tmp_s = normalized_linear_prod(b, &ec, b.examples[loc]);
+                    example* kprod_ec = &calloc_or_throw<example>();
+                    diag_kronecker_product_test(ec, *b.examples[loc], *kprod_ec, b.oas);
+                    kprod_ec->l.simple = {FLT_MAX, 0., tmp_s};
+                    base.predict(*kprod_ec, b.max_routers);
+                    score = kprod_ec->partial_prediction;
+                    free_example(kprod_ec);
+                }
+                else
+                    score = normalized_linear_prod(b, &ec, b.examples[loc]);
+                
+                if (score > max_score){
+                    max_score = score;
+                    max_pos = (int64_t)loc;
+                }
+            }
+            return max_pos;
+        }
+        else
+            return -1;
+    }
+
+    //for any two examples, use number of overlap labels to indicate the similarity between these two examples. 
+    float get_overlap_from_two_examples(example& ec1, example& ec2){
+        int num_overlap = over_lap(ec1.l.multilabels.label_v, ec2.l.multilabels.label_v);
+        return num_overlap*1.f;
+    }
+
+    //we use F1 score as the reward signal 
+    float F1_score_for_two_examples(example& ec1, example& ec2){
+        float num_overlaps = get_overlap_from_two_examples(ec1, ec2);
+	    float v1 = num_overlaps/(1e-7+ec1.l.multilabels.label_v.size()*1.);
+        float v2 = num_overlaps/(1e-7+ec2.l.multilabels.label_v.size()*1.);
+        if (num_overlaps == 0.f)
+            return 0.f;
+        else
+	    //return v2; //only precision
+            return 2.*(v1*v2/(v1+v2));
+    }
+
+    void predict(memory_tree& b, single_learner& base, example& test_ec)
+    {
+        example& ec = calloc_or_throw<example>();
+        copy_example_data(&ec, &test_ec, b.oas);
+        remove_repeat_features_in_ec(ec);
+	//example& ec = test_ec;
+     
+
+        MULTICLASS::label_t mc;
+        uint32_t save_multi_pred;
+        MULTILABEL::labels multilabels;
+        MULTILABEL::labels preds;
+        if (b.oas == false){
+            mc = ec.l.multi;
```

Comment:
The predict pipeline really shouldn't be touching labels.

Reply:
I was just trying to temporally save the label information, and after base.predict is called, I resigned it back (this seems the pattern that I observed in other files like recall tree and Lom tree?)

Reply:
It should not be necessary, because predict does not look at or use label information.

Reply:
I did some quick test, if I don't save/reload the label information, then the original value contained in ec.l.multi.label is lost after the base.predict call here..

Reply:
Ok.   Something elsewhere needs refactoring.

Reply:
I guess I will mark it as resolved now? I couldn't figure out what's going on...(I feel that I don't understand union quite well enough, and I recall this issue costed me a lot of time during debugging when I was writing the code, so for safety reason I later on just added save/load every time before and after I call base.predict )

Reply:
ok.